zenodo/modules/metrics/api.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
+# Copyright (C) 2016, 2017 CERN.
+#
+# Zenodo is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Zenodo is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Zenodo; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Zenodo Metrics API."""
+
+from __future__ import absolute_import
+
+from elasticsearch_dsl import A, Q
+from invenio_accounts.models import User
+from invenio_communities.models import Community
+from invenio_files_rest.models import FileInstance
+from invenio_search.api import RecordsSearch
+from .proxies import current_metrics
+
+
+class ZenodoMetric(object):
+    """API class for Zenodo Metrics."""
+
+    @staticmethod
+    def get_users():
+        """Get number of unique zenodo users."""
+        return 1
+
+    @staticmethod
+    def get_visitors():
+        """Get number of unique zenodo users."""
+        # Query: '"aggs":{"visitors_count":{"cardinality":{' \
+        #                '"field":"visitor_id","precision_threshold":' \
+        #                '40000}}}'
+        # search = RecordsSearch(index='events-stats-*') \
+        #     .aggs.bucket('visitors_count', 'value_count',
+        #                  field='visitor_id')
+        # return search.count()
```

Comment:
```suggestion
        # Query: '"aggs":{"visitors_count":{"cardinality":{' \
        #                '"field":"visitor_id","precision_threshold":' \
        #                '40000}}}'
        search = Search(
            using=current_search_client,
            index='events-stats-*'
        ).filter(
            'range', ...,
        ).aggs.metric(
            'visitors_count', 'cardinality', field='visitor_id'
        )
        result = search.execute()
        return result.aggregations.visitors_count.value
```

- I think you're right about the `RecordSearch` class, it probably messes things up. You can use the base ElasticSearch-DSL `Search` class as done [here](https://github.com/zenodo/zenodo/blob/9e044a035a4b66d75949c609f29794c44e2b739e/zenodo/modules/stats/exporters.py#L63-L65). Basically you have to pass the `using=<client>` parameter, and apply some index prefixing (i.e. to make `events-stats-*` -> `zenodo-events-stats-*`).
- We also need to filter results in the specific time range (i.e. starting from 2021-01-01), but see the next point.
- For some inspiration for using the `cardinality` metric aggregation over a specific time range see [this example](https://github.com/inveniosoftware/invenio-stats/blob/master/invenio_stats/utils.py#L49-L61).